solidity/ecdsa/contracts/WalletFactory.sol

```diff
@@ -0,0 +1,306 @@
+// SPDX-License-Identifier: MIT
+//
+// ▓▓▌ ▓▓ ▐▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▄
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓    ▓▓▓▓▓▓▓▀    ▐▓▓▓▓▓▓    ▐▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▄▄▓▓▓▓▓▓▓▀      ▐▓▓▓▓▓▓▄▄▄▄         ▓▓▓▓▓▓▄▄▄▄         ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▓▓▓▓▓▓▓▀        ▐▓▓▓▓▓▓▓▓▓▓         ▓▓▓▓▓▓▓▓▓▓         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▀▀▓▓▓▓▓▓▄       ▐▓▓▓▓▓▓▀▀▀▀         ▓▓▓▓▓▓▀▀▀▀         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▀
+//   ▓▓▓▓▓▓   ▀▓▓▓▓▓▓▄     ▐▓▓▓▓▓▓     ▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌
+// ▓▓▓▓▓▓▓▓▓▓ █▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+//
+//                           Trust math, not hardware.
+
+pragma solidity ^0.8.9;
+
+import "./libraries/DKG.sol";
+import "./Wallet.sol";
+import "./DKGValidator.sol";
+import "@keep-network/sortition-pools/contracts/SortitionPool.sol";
+import "@thesis/solidity-contracts/contracts/clone/CloneFactory.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+/// TODO: Add a dependency to `threshold-network/solidity-contracts` and use
+/// IStaking interface from there.
+interface IWalletStaking {
+    function eligibleStake(address operator, address operatorContract)
+        external
+        view
+        returns (uint256);
+}
+
+contract WalletFactory is CloneFactory, Ownable {
+    using DKG for DKG.Data;
+
+    // Libraries data storages
+    DKG.Data internal dkg;
+
+    Wallet[] public wallets;
```

Comment:
Will there be a case when we need to iterate this array? If no, how about we change the array to a map? I did some experiments map vs. array:

```
Wallet[] public wallets;
mapping(bytes32 => Wallet) public walletsByPubKeyHash; // alternative with mapping

(...)
wallets.push(wallet) // gas: 44297
walletsByPubKeyHash[wallet.publicKeyHash()] = wallet; // gas: 23243

wallets.pop(); // gas: 6147
delete walletsByPubKeyHash[keccak256(dkgResult.groupPubKey)]; // gas: 5466
```

It is not that significant when removing a wallet especially it won't be that frequent, but when adding we have ~21k of difference in favor of mapping. What do you think? btw. sorry for the late comment, I could've proposed this earlier.